src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -1258,18 +1276,84 @@ private void LoggingEventProcessor(object loggingEvent)
+
             if (loggingEvent is BuildEventArgs)
             {
-                RouteBuildEvent((BuildEventArgs)loggingEvent);
+                buildEventArgs = (BuildEventArgs)loggingEvent;
             }
             else if (loggingEvent is KeyValuePair<int, BuildEventArgs>)
             {
-                RouteBuildEvent((KeyValuePair<int, BuildEventArgs>)loggingEvent);
+                buildEventArgs = ((KeyValuePair<int, BuildEventArgs>)loggingEvent).Value;
             }
             else
             {
                 ErrorUtilities.VerifyThrow(false, "Unknown logging item in queue:" + loggingEvent.GetType().FullName);
             }
+
+            if (buildEventArgs is BuildWarningEventArgs warningEvent)
+            {
+                if (ShouldTreatWarningAsMessage(warningEvent))
+                {
+                    loggingEvent = new BuildMessageEventArgs(
+                        warningEvent.Subcategory,
+                        warningEvent.Code,
+                        warningEvent.File,
+                        warningEvent.LineNumber,
+                        warningEvent.ColumnNumber,
+                        warningEvent.EndLineNumber,
+                        warningEvent.EndColumnNumber,
+                        warningEvent.Message,
+                        warningEvent.HelpKeyword,
+                        warningEvent.SenderName,
+                        MessageImportance.Low,
+                        warningEvent.Timestamp)
+                    {
+                        BuildEventContext = warningEvent.BuildEventContext,
+                        ProjectFile = warningEvent.ProjectFile,
+                    };
+                }
+                else if (ShouldTreatWarningAsError(warningEvent))
+                {
+                    loggingEvent = new BuildErrorEventArgs(
+                        warningEvent.Subcategory,
+                        warningEvent.Code,
+                        warningEvent.File,
+                        warningEvent.LineNumber,
+                        warningEvent.ColumnNumber,
+                        warningEvent.EndLineNumber,
+                        warningEvent.EndColumnNumber,
+                        warningEvent.Message,
+                        warningEvent.HelpKeyword,
+                        warningEvent.SenderName,
+                        warningEvent.Timestamp)
+                    {
+                        BuildEventContext = warningEvent.BuildEventContext,
+                        ProjectFile = warningEvent.ProjectFile,
+                    };
+                }
+            }
+
+            if (loggingEvent is BuildErrorEventArgs errorEvent)
+            {
+                // Keep track of build submissions that have logged errors.  If there is no build context, add BuildEventContext.InvalidSubmissionId.
+                _buildSubmissionIdsThatHaveLoggedErrors.Add(errorEvent.BuildEventContext?.SubmissionId ?? BuildEventContext.InvalidSubmissionId);
```

Comment:
>_buildSubmissionIdsThatHaveLoggedErrors [](start = 16, length = 39)

I realize now that `_buildSubmissionIdsThatHaveLoggedErrors` is per node, so hopefully its value is not required in the main node. This problem existed with the old implementation too.

Reply:
Its set by all nodes but only read by the main node.